appsrc/dcfile/dclutdmp.man
.TH DCLUTDMP 1 "6 March 2007" "DICOM PS3" "DICOM PS3 - dump LUT contents"
.SH NAME
dclutdmp \- ACR/NEMA DICOM PS3 ... DICOM PS3 - dump LUT contents
.SH SYNOPSIS
.HP 10
.B dclutdmp
[
.B \-ancreate
]
[
.B \-min value
]
[
.B \-max value
]
[
.B \-v|verbose
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dclutdmp
reads the named dicom input file and dumps the contents of any VOI LUT
present (as decimal values).
.SH OPTIONS
The descriptive attributes go to standard error.
.PP
The tab delimited input vs. output value list goes to stdout.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-ancreate
.RS
Display item, end item and end sequence delimiter tags in a form suitable to feed into ancreate.
.RE
.TP
.B \-min value
.TP
.B \-max value
.RS
Specify the first and last input values to dump ... if these are greater than
the specified input range of the actual LUT, then the first and last output
values are replicated as per the standard. The values may also be less than
the actual LUT range. This option makes it easier to
concatenate two LUTs specified over different input values for the purposes
of comparison.
.RE
.TP
.B \-verbose
.RS
Dump the complete list of DICOM attributes present in the input to stderr.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR ancreate(1) ,
.BR andump(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dclutdmp.cc
#include "attrmxls.h"
#include "attrval.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"

#include "errclass.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool showfilename=options.get("filename");

	unsigned input_min = 0x0000;
	unsigned input_max = 0xffff;
	bool minpresent=options.get("min",input_min);
	bool maxpresent=options.get("max",input_max);

	bool ancreate=options.get("ancreate");
	
	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-ancreate]"
			<< " [-min value]"
			<< " [-max value]"
			<< " [-v|-verbose]"
			<< " [-filename]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		return 1;
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	TextOutputStream  log(cerr);
	
	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		//return 1;	// Press on regardless ...
	}


	Attribute *aVOILUTSequence = list[TagFromName(VOILUTSequence)];
	int nVOILUT=0;
	if (aVOILUTSequence) {
		if (ancreate) {
			log << "(0x0028,0x3010) SQ VOI LUT Sequence VR=<SQ> VL=<0xffffffff> []" << endl;
		}
		else {
			log << "VOILUTSequence" << endl;
		}
		AttributeList **items;
		nVOILUT=aVOILUTSequence->getLists(&items);
		if (!ancreate) {
			log << "Contains " << dec << nVOILUT << " items" << endl;
		}
		int item;
		for (item=0; item<nVOILUT; ++item) {
			if (ancreate) {
				log << "%item" << endl;
			}
			else {
				log << "\t Item " << dec << item << endl;
			}
			Assert(items[item]);
			Attribute *aLUTExplanation = items[item]->operator[](TagFromName(LUTExplanation));
			if (aLUTExplanation) {
				const char *vLUTExplanation=AttributeValue(aLUTExplanation);
				if (!ancreate) {
					log << "\t\t LUTExplanation  = " << vLUTExplanation << endl;
				}
			}
			
			Attribute *aLUTDescriptor=items[item]->operator[](TagFromName(LUTDescriptor));
			if (!aLUTDescriptor) {
				log << "\t\t Missing LUT Descriptor Attribute" << endl;
			}
			else {
				if (ancreate) {
					aLUTDescriptor->write(log,list.getDictionary(),false);
					log << endl;
					aLUTExplanation->write(log,list.getDictionary(),false);			// must go after descriptor
					log << endl;
				}
				unsigned vm=aLUTDescriptor->getVM();
				if (vm != 3) {
					log << "\t\t Wrong VM for LUT Descriptor = " << dec << vm << endl;
				}
				else {
					Uint16 VOILUT_first;
					Uint16 VOILUT_number;
					Uint16 VOILUT_depth;
					Assert(aLUTDescriptor->getValue(0,VOILUT_number));
					Assert(aLUTDescriptor->getValue(1,VOILUT_first));
					Assert(aLUTDescriptor->getValue(2,VOILUT_depth));
					if (!ancreate) {
						log << "\t\t first  = " << dec << VOILUT_first << endl
							<< "\t\t number = " << dec << VOILUT_number << endl
							<< "\t\t depth  = " << dec << VOILUT_depth << endl;
					}
					Attribute *aLUTData=items[item]->operator[](TagFromName(LUTData));
					if (!aLUTData) {
						log << "\t\t Missing LUT Data Attribute" << endl;
					}
					else {
						if (ancreate) {
							//aLUTData->write(log,list.getDictionary(),false);
							//log << endl;
							log << "(0x0028,0x3006) XO LUT Data      VR=<OW>   VL=<" << hex << (VOILUT_number*2) << dec << "> [";
						}
						//else {
						{
							const char *prefix = "";
							int j;
							Uint16 VOILUT_value;
							Uint32 nLUTData;
							if (aLUTData->isOtherWordNonPixel()) {
//cerr << "checking > 8 bit OW" << endl;
								const Uint16 *vLUTData = NULL;
								if (aLUTData->getValue(vLUTData,nLUTData)) {
//cerr << "got value" << endl;
									if (!ancreate) {
										if (minpresent && VOILUT_number && nLUTData) {
											for (j=input_min; j < VOILUT_first; ++j) {
												cout << dec << j << "\t" << VOILUT_value << endl;
											}
										}
									}
									int i;
									for (i=0,j=VOILUT_first; i < VOILUT_number; ++i,++j) {
										VOILUT_value = vLUTData[i];
										if (ancreate) {
											log << prefix << hex << VOILUT_value << dec << "";
											prefix=",";
										}
										else {
											if (j >= input_min && j <= input_max) {
												cout << dec << j << "\t" << VOILUT_value << endl;
											}
										}
									}
								}
								else {
									log << "\t\t Cannot get OW LUT Data" << endl;
								}
							}
							else if (aLUTData->isNumericBinary()) {		// e.g. US VR
//cerr << "checking " << aLUTData->getVR() << endl;
								nLUTData = aLUTData->getVM();
								if (minpresent && VOILUT_number && nLUTData && aLUTData->getValue(0,VOILUT_value)) {
									if (!ancreate) {
										for (j=input_min; j < VOILUT_first; ++j) {
											cout << dec << j << "\t" << VOILUT_value << endl;
										}
									}
								}
								int i;
								for (i=0,j=VOILUT_first; i < VOILUT_number; ++i,++j) {
									if (aLUTData->getValue(i,VOILUT_value)) {
										if (ancreate) {
											log << prefix << hex << VOILUT_value << dec << "";
											prefix=",";
										}
										else {
											if (j >= input_min && j <= input_max) {
												cout << dec << j << "\t" << VOILUT_value << endl;
											}
										}
									}
									else {
										log << "\t\t Premature End of LUT Data" << endl;
										break;
									}
								}
							}
							// at this point, for either encoding, VOILUT_value will contain the last value
							if (!ancreate) {
								if (maxpresent && VOILUT_number) {
									for (j=VOILUT_first+VOILUT_number; j <= input_max; ++j) {
										cout << dec << j << "\t" << VOILUT_value << endl;
									}
								}
							}
						}
						if (ancreate) {
							log << "]" << endl;
						}
					}
				}
			}
			if (ancreate) {
				log << "%enditem" << endl;
			}
		}
		if (ancreate) {
			log << "%endseq" << endl;
		}
	}

	return 0;
}

